### PR TITLE
Basic support to send emails [#472]

### PR DESCRIPTION
**Pull request by np**
No support yet for the HTML optional body.

### CODE REVIEW

**Comment by np** [comment]
This ready for a first round of reviews. I still have some issues with my setup, so I've not ran the tests myself.

File: thentos-core/src/Thentos/Backend/Api/Simple.hs [hunk 1]
```diff
@@ -161,6 +163,11 @@ thentosServiceSession =
   :<|> endServiceSession
 
 
+-- * send email
+
+type SendEmail =
+    ReqBody '[JSON] SendEmailRequest :> Post '[JSON] ()
+
 -- * servant docs
```
Comment (fisx) [inline thentos-core/src/Thentos/Backend/Api/Simple.hs]: There are some layout style rules for haddock sections:

https://github.com/liqd/thentos/blob/master/docs/styleguide.md#layout

(There is also the rule that if you don't like anything in the style guide, you can write a PR.  And regarding haddock section headings, I'm not sure they even make sense in the presence of explicit export list.  I think in that case they should go to the export list, so this rule needs some work.  Anyway, this is what it is now...)

File: thentos-core/src/Thentos/Types.hs [hunk 2]
```diff
@@ -22,7 +22,7 @@ module Thentos.Types
     , PasswordResetToken(..)
     , PasswordResetRequest(..)
     , UserFormData(..)
-    , UserCreationRequest(..)
+    , UserCreationRequest(..), emailRecipient, emailSubject, emailPlainTextBody
```
Comment (fisx) [inline thentos-core/src/Thentos/Types.hs]: Are these in the wrong line?

File: thentos-core/src/Thentos/Types.hs [hunk 3]
```diff
@@ -507,6 +509,22 @@ instance ToJSON ByUserOrServiceId where
     toJSON (ByService i k) = Aeson.object [ "service" .= Aeson.toJSON (i, k)]
 
 
+-- * send email
+
+-- TODO: Service ID? Pseudonyms? From address?
```
Comment (fisx) [inline thentos-core/src/Thentos/Types.hs]: For from address, you can use this from `Thentos.Config`:

```haskell
type SmtpConfig' =
      Maybe ("sender_name"    :> ST)  -- FIXME: use more specific type 'Network.Mail.Mime.Address'
  :*>       ("sender_address" :> ST)
...
```

Comment (fisx) [inline thentos-core/src/Thentos/Types.hs]: Service ID is not required for now.

Pseudonyms: yes, that would be nice.  Do you need details on this?  Essentially, they are a services handle for a user, and need to be looked up, mapped to the user, and then lead to the email address.

File: thentos-core/src/Thentos/Types.hs (hunk 2 above)
Comment (np) [inline thentos-core/src/Thentos/Types.hs]: Yes, it's a mistake.

File: thentos-core/src/Thentos/Types.hs (hunk 3 above)
Comment (np) [inline thentos-core/src/Thentos/Types.hs]: I'm using the smtp config already, as I'm using the same function calls.

Comment (fisx) [inline thentos-core/src/Thentos/Types.hs]: Ah, you were asking if we want the sender to be able to change that: no, default sender from the config is fine.

File: thentos-core/src/Thentos/Action.hs [hunk 4]
```diff
@@ -613,6 +615,16 @@ getServiceSessionMetadata :: ServiceSessionToken -> Action e s ServiceSessionMet
 getServiceSessionMetadata tok = (^. srvSessMetadata) <$> lookupServiceSession tok
 
 
+-- * send emails
+
+-- | Send an email. Only a privileged IP is allowed to use this endpoint.
+sendEmail :: SendEmailRequest -> Action e s ()
+sendEmail req = do
+    void hasPrivilegedIP
```
Comment (fisx) [inline thentos-core/src/Thentos/Action.hs]: I wonder if `hasPrivilegedIP` should be changed to make this `void` unnecessary?  (Can't be bothered to check myself, not necessarily part of this PR, just wondering...)

**Comment by fisx** [comment]
Notes from the phone call:

- we want to be able to send to lists of recipients.
- "recipient" can be "persona", "email", "group".  how those are distinguished still needs to be designed.
- templating: like in the activate_user end point.
- mass mailing: not an issue yet.  we can trust privileged IPs.

afterthough: ./docs/messaging.md still contains a few fixmes, and i think most of them are things that can be designed using nothing but common sense and general knowledge.  @np, could you add that either here or in a separate PR?

**Comment by ChristianSi** [comment]
> afterthough: ./docs/messaging.md still contains a few fixmes, and i think most of them are things that can be designed using nothing but common sense and general knowledge. @np, could you add that either here or in a separate PR?

There is one FIXME about the return value of the new endpoint. I proposed

    {"data": {}}

which, with some hindsight, doesn't make any sense. In your PR, the return value is `Post '[JSON] ()` which (I think) corresponds HTTP status code `204 No Content` and no body. I think this is very reasonable and the messaging.md should be adapted to document it.

**Comment by fisx** [comment]
Clarification:

> - "recipient" can be "persona", "email", "group".  how those are distinguished still needs to be designed.

I meant `PersonaId` here.  This is available to the service and the default way to refer to a persona.

**Comment by fisx** [comment]
Nice to have: send email to `ServiceGroup`.

There are `Group`s of users (organising things in thentos) and `ServiceGroup`s of personas (organising the same things on the service side, i.e. adhocracy).  See ./docs/terminology.md for more details.

The service may want to send email to all (transitive) members of a service group.  For this to work, I think there may be data types, transactions and actions missing involving the use (not creation and assignment) of service groups.  Currently, we have:

```sh
grep persona_groups thentos-core/schema/schema.sql thentos-core/src/Thentos/Transaction.hs
grep group_tree thentos-core/schema/schema.sql thentos-core/src/Thentos/Transaction.hs
```

Please ping me if you want me to elaborate on this.

**Comment by fisx** [comment]
Updated earlier comment.  Now:

> I meant ~~`PersonaId`~~ `personaExternalUrl`

**Comment by np** [comment]
Hum, thanks to git rebase the date of this commit should have been now and not yesterday.

**Comment by np** [comment]
This is ready for another round of review @fisx @ChristianSi

**Comment by fisx** [comment]
Looks good.  Just a few comments.

**Comment by np** [comment]
While my last commit ec0071a might be better for incremental review, I'm going to rebase onto master to make it a proper PR.

**Comment by np** [comment]
Actually I rebased but preserved the separate commits at least for now.

**Comment by fisx** [comment]
thanks for keeping the individual commits!  losing them would have been annoying.

generally speaking, i was asked not to rebase in between incremental reviews, and now i see why :-).  i'll assume though that the old commits haven't changed and will just read the last one.

**Comment by fisx** [comment]
I made two small commits.  Could you check if you have any objections and otherwise merge?

**Comment by np** [comment]
LGTM

**Pull request merged**